app/javascript/hyacinth_v1/graphql/fileUploads.js

```diff
@@ -1,9 +1,14 @@
 import * as ActiveStorage from 'activestorage';
 
-const uploadFile = (file, url) => {
+const uploadFile = (file, url, onProgress) => {
   const delegate = {
-    directUploadWillCreateBlobWithXHR: (data) => { console.debug(data); },
-    directUploadWillStoreFileWithXHR: (data) => { console.debug(data); },
+    directUploadWillCreateBlobWithXHR: (request) => { console.debug(request); },
+    directUploadWillStoreFileWithXHR: (request) => {
+      request.upload.addEventListener('progress', (event) => {
```

Comment:
faint memory ringing - is there a memory issue with disposal of these event listeners?

Reply:
This is a good concern, and I've been doing some reading and it seems that it's not a problem in today's modern browsers -- with plain JavaScript, at least.  Use of jQuery may increase the chance of memory leaks because of the different way that jQuery binds events (we'd want to look more into that case), but fortunately we're not using jQuery here.  React does some fancy things behind the scenes with the way it manages the DOM, but since React promotes the use of `onClick` (rather than the usual `onclick`) and hasn't required explicit cleanup of those event handlers, I'm guessing that pattern also avoids memory leaks.

One other thing that makes this particular case even better for garbage collection is that the event listener is bound to a request object rather than a DOM element.  One of the major reasons for memory leaks in older browsers (especially IE) was a garbage collection issue where event listeners held onto references to DOM elements, which caused issues with cleanup of those elements, and I believe this led to circular reference cleanup issues.

Mentioning @cgalarza here just as an FYI, since this is good for all of us to know about going forward.

Reply:
I guess another minor performance point: It's kind of extravagant to have 3 inline lambdas with the same body, isn't it?

Reply:
It is, though I think not doing so makes it harder to pass `onProgress` to the `addEventListener` inner context (unless we can do something with `bind`, maybe?).  Do you have a preferred refactor in mind?

Here's a slightly cleaned up version, at least.  I removed some stuff that appeared to be unused.  What do you think?

```
const uploadFile = (file, url, onProgress) => {
  const delegate = {
    directUploadWillStoreFileWithXHR: (request) => {
      request.upload.addEventListener('progress', (event) => {
        onProgress(Math.round(event.loaded * 100 / event.total));
      });
    },
  };
  const directUpload = new ActiveStorage.DirectUpload(file, url, delegate);
  return new Promise((resolve, reject) => {
    directUpload.create((error, props) => {
      if (error) {
        reject(error);
      } else {
        resolve(props);
      }
    });
  });
};
```